internal/privatestate/data.go

```diff
@@ -0,0 +1,178 @@
+			fmt.Sprintf("An error was encountered when encoding private state: %s.\n\n"+
+				"This is always a problem with Terraform or terraform-plugin-framework. Please report this to the provider developer.", err),
+		)
+
+		return nil, diags
+	}
+
+	return bytes, diags
+}
+
+// NewData creates a new Data based on the given slice of bytes.
+// It must be a JSON encoded slice of bytes, that is map[string][]byte.
+func NewData(ctx context.Context, data []byte) (Data, diag.Diagnostics) {
+	var (
+		dataMap map[string][]byte
+		diags   diag.Diagnostics
+	)
+
+	err := json.Unmarshal(data, &dataMap)
+	if err != nil {
+		diags.AddError(
+			"Error Decoding Private State",
+			fmt.Sprintf("An error was encountered when decoding private state: %s.\n\n"+
+				"This is always a problem with Terraform or terraform-plugin-framework. Please report this to the provider developer.", err),
+		)
+
+		return Data{}, diags
+	}
+
+	output := Data{
+		Framework: make(map[string][]byte),
+		Provider:  make(map[string][]byte),
+	}
+
+	for k, v := range dataMap {
+		if isInvalidProviderDataKey(ctx, k) {
+			output.Framework[k] = v
+			continue
+		}
+
+		output.Provider[k] = v
+	}
+
+	return output, diags
+}
+
+// ProviderData contains private state data for provider usage.
+type ProviderData map[string][]byte
+
+// GetKey returns the private state data associated with the given key.
+//
+// If the key is reserved for framework usage, an error diagnostic
+// is returned. If the key is valid, but private state data is not found,
+// nil is returned.
+//
+// The naming of keys only matters in context of a single resource,
+// however care should be taken that any historical keys are not reused
+// without accounting for older resource instances that may still have
+// older data at the key.
+func (d ProviderData) GetKey(ctx context.Context, key string) ([]byte, diag.Diagnostics) {
+	diags := ValidateProviderDataKey(ctx, key)
+
+	if diags.HasError() {
+		return nil, diags
+	}
+
+	value, ok := d[key]
+
+	if !ok {
+		return nil, nil
+	}
+
+	return value, nil
+}
+
+// SetKey sets the private state data at the given key.
+//
+// If the key is reserved for framework usage, an error diagnostic
+// is returned. The data must be valid JSON and UTF-8 safe or an error
+// diagnostic is returned.
+//
+// The naming of keys only matters in context of a single resource,
+// however care should be taken that any historical keys are not reused
+// without accounting for older resource instances that may still have
+// older data at the key.
+func (d ProviderData) SetKey(ctx context.Context, key string, value []byte) diag.Diagnostics {
+	diags := ValidateProviderDataKey(ctx, key)
+
+	if diags.HasError() {
+		return diags
+	}
+
+	if !utf8.Valid(value) {
+		diags.AddError("UTF-8 Invalid",
+			"Values stored in private state must be valid UTF-8\n\n"+
+				"This is always a problem with Terraform or terraform-plugin-framework. Please report this to the provider developer.",
```

Comment:
Since this data provided from the provider developer in their codebase, should this messaging be adjusted? It might also be good to include the key that is causing the error and maybe a log message with the value for troubleshooting.

Reply:
We could also consider adding the value to the error diagnostic, however I think we should be conservative with the potential to output sensitive data.

Reply:
D'oh! Good point.